### PR TITLE
[Git] Unbreak the addin

### PR DESCRIPTION
**Pull request by Therzok**
We need to offset based on hunk inserted/removed lengths

Fixes VSTS 978853 - [VersionControl] GitRepository: Error retrieving history System.ArgumentException: Offset and length were out of bounds
Fixes VSTS #977874 - Renaming file makes intellisense disapear when project has .git
Fixes VSTS #787121 - [Watson] Git commands are run on the UI thread, causing the UI to hang

### CODE REVIEW

**Comment by Therzok** [comment]
@monojenkins backport release-8.3

**Comment by Therzok** [comment]
@monojenkins backport release-8.3

**Comment by Therzok** [comment]
@monojenkins backport release-8.3

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlFileSystemExtension.cs [hunk 1]
```diff
@@ -103,18 +100,15 @@ public override void MoveDirectory (FilePath sourcePath, FilePath destPath)
...
-					try {
-						await srcRepo.MoveDirectoryAsync (sourcePath, destPath, true, monitor);
-					} catch (OperationCanceledException) {
-						return;
-					} catch (Exception e) {
-						LoggingService.LogError ("Error while moving directory.", e);
-					} finally {
-						Monitor.Exit (moveDirectoryLock);
+				try {
+					srcRepo.MoveDirectoryAsync (sourcePath, destPath, true, monitor).Wait();
```
Comment (Therzok) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlFileSystemExtension.cs]: I don't like these .Waits(), but there's nothing else we can do.

**Comment by Therzok** [comment]
@monojenkins backport release-8.3

**Comment by Therzok** [comment]
@monojenkins backport release-8.3

**Comment by Therzok** [comment]
Also, fixes VSTS #978271 - [Feedback] [Git Version Control] Crash when applying a stash

File: main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/GitConfigurationDialog.cs [hunk 2]
```diff
@@ -452,15 +452,15 @@ protected async void OnButtonFetchClicked (object sender, EventArgs e)
 				return;
 
 			var monitor = VersionControlService.GetProgressMonitor (GettextCatalog.GetString ("Fetching remote..."));
-			await System.Threading.Tasks.Task.Run (() => {
+			await System.Threading.Tasks.Task.Run ((Func<System.Threading.Tasks.Task>)(async () => {
```
Comment (slluis) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/GitConfigurationDialog.cs]: Do we need the Task.Run if Fetch is now async?

Comment (Therzok) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/GitConfigurationDialog.cs]: No, we don't! Good catch

Comment (Therzok) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl.Git/MonoDevelop.VersionControl.Git/GitConfigurationDialog.cs]: Fixed.

**Comment by Therzok** [comment]
@monojenkins backport release-8.3

**Comment by Therzok** [comment]
@mrward can add a comment. Basically the diff indices are relative to the base directory, not the working directory. So we need to offset the indices based on how many lives were inserted/removed. New test didn't pass before.

**Comment by mrward** [comment]
@Therzok I am sure it is fine. No need for a comment in the code.

**Pull request merged**